py/gpu_specter/test/parity.py

```diff
@@ -0,0 +1,140 @@
+    ax.set_title('abs( (f_a - f_b)/sqrt(var_a + var_b) )')
+    fig.colorbar(im, ax=ax, extend='both')
+    plt.savefig(output + '-2D.png', bbox_inches='tight', dpi=100)
+
+def plot_diff_1d(diff, wave, output):
+    fig, ax = plt.subplots(figsize=(8, 6))
+    # pick some arbitrary spectra
+    for ispec in (np.arange(20)*25 + 7):
+        ax.plot(wave, 1e2*diff[ispec] + ispec, marker='.', markersize=1, lw=0)
+
+    ax.set_ylabel('1e2*diff (offset by spectrum index)')
+    ax.set_xlabel('wavelength')
+    plt.savefig(output + '-1D.png', bbox_inches='tight', dpi=100)
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Process some integers.')
+    parser.add_argument('-a', '--filename-a', type=str, required=True,
+                        help='filename a')
+    parser.add_argument('-b', '--filename-b', type=str, required=True,
+                        help='filename b')
+    parser.add_argument('-o', '--output', type=str, default=None,
+                        help='output filename base')
+    args = parser.parse_args()
+
+    # silence warning caused by dividing by ivar==0
+    np.seterr(divide='ignore', invalid='ignore')
+
+    a = read_frame(args.filename_a)
+    b = read_frame(args.filename_b)
+
+    assert a['wave'].shape == b['wave'].shape
+    assert a['flux'].shape == b['flux'].shape
+    assert a['ivar'].shape == b['ivar'].shape
+
+    # compare wave
+    assert np.all(a['wave'] == b['wave'])
+
+    nspec = a['flux'].shape[0]
+    wave = a['wave']
+    wmin, wmax = wave[0], wave[-1]
+    nwave = len(wave)
+
+    template = '{label:>10}: {fraction:7.2%}'
+
+    # compare flux
+    print('(f_a, f_b):')
+    isclose = np.average(np.isclose(a['flux'], b['flux']))
+    print(template.format(label='isclose', fraction=isclose))
+
+    print('(f_a - f_b)/sqrt(var_a + var_b):')
+    diff = (a['flux'] - b['flux']) / np.sqrt(1.0/a['ivar'] + 1.0/b['ivar'])
+    tstart, tstop = -5, 0
+    thresholds = np.logspace(tstart, tstop, (tstop - tstart) + 1, endpoint=True)
+    for threshold in thresholds:
+        fraction = np.average(np.abs(diff).ravel() < threshold)
+        print(template.format(label=threshold, fraction=fraction))
+
+
+    if args.output is not None:
+        plot_diff_2d(diff, wave, args.output)
+        plot_diff_1d(diff, wave, args.output)
+
+    # compare ivar
+    print('(ivar_a, ivar_b):')
+    isclose = np.average(np.isclose(a['ivar'], b['ivar']))
+    print(template.format(label='isclose', fraction=isclose))
+
+    print('(sigma_a - sigma_b)/sqrt(var_a + var_b):')
+    avar = 1.0/a['ivar']
+    bvar = 1.0/b['ivar']
+    diff = (np.sqrt(avar) - np.sqrt(bvar)) / np.sqrt(avar + bvar)
+    tstart, tstop = -5, 0
+    thresholds = np.logspace(tstart, tstop, (tstop - tstart) + 1, endpoint=True)
+    for threshold in thresholds:
+        fraction = np.average(np.abs(diff).ravel() < threshold)
+        print(template.format(label=threshold, fraction=fraction))
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
For the record: our standard DESI style is to put anything that is expected to be run from the command line in bin/ (and without a .py extension), and have that wrap library code to be imported in py/ and avoid `if __name__ == "__main__":` stuff under py/.  However, that can result in polluting the $PATH namespace with a bunch of specialized one-off scripts, so I'm sympathetic to having specialized scripts like this "hidden" within py/ as long as the procedures are documented.

i.e. this is fine as is, but commenting on it stylistically as we consider where future stuff goes and how this should be organized when it is ported back to specter and desispec.

Reply:
Moving to bin/ at the expense of polluting $PATH seems like a reasonable tradeoff for now.

	renamed:    py/gpu_specter/test/parity.py -> bin/compare-frame
	renamed:    py/gpu_specter/test/README.md -> doc/testing.md